README.md
# rgeoprofile: Geographic Profiling Methods for Serial Crime Analysis <a><img src='../master/images/rgeoprofile-hex.png' align="right" height="139" /></a>

<!-- badges: start -->
![CRAN Version](https://www.r-pkg.org/badges/version-ago/rgeoprofile)
![CRAN Downloads](https://cranlogs.r-pkg.org/badges/last-month/rgeoprofile)
![Build Status](https://travis-ci.org/JSSpaulding/rgeoprofile.svg?branch=master)
<!-- badges: end -->

## Overview

**rgeoprofile** is a package containing various functions for the analysis of serial crime incidents. The package implements algorithms for the geographical profiling of serial incidents in attempt to prioritize the area in which the anchor point or home base of the perpetrator is located. These functions include centrographic methods (center of minimum distance, geometric and harmonic means) and distance decay functions in the following forms: Rossmo's formula, linear, negative exponential (CrimeStat and Dragnet, with/out plateau and buffer), normal, lognormal, and truncated negative exponential.
	
**Keywords**: Crime Analysis, Geographic Profiling, Distance Decay, Open Source

## Getting Started

These instructions will allow any user to install the R package on your local machine for development and testing purposes. 

### Installing

Begin by installing the package from the GitHub repository:

```
devtools::install_github('JSSpaulding/rgeoprofile')
```

After installation, load and attach the package

```
library(rgeoprofile)
```

The following will call example serial crime incident data from within the package for the Boston Strangler (Albert DeSalvo) case. This data structure includes the following columns: victim name, age, date of incident, location (street address), latitude, and longitude.

```
desalvo <- data.frame(rgeoprofile:::boston_strangler)
```

## Distance Decay Model Usage

The following examples illustrate the usage of the various distance decay functions within the **rgeoprofile** package.

### cgt_profile

The *cgt_profile* function applies the criminal geographic targeting (CGT) model for serial crime analysis developed by DK Rossmo for geographic profiling and prediction of perpetrator anchor point/home base. The CGT distance decay function assumes a buffer zone around the incidents where the likelihood of perpetrator anchor point increases to a peak likelihood before decaying as distance further increases. This model also utilizes Manhattan distance for the likelihood calculation.

When plotted, the likelihood of the perpetrator's anchor point is illustrated by the following:

![](../master/images/Rossmo_ddf.png)

```
desalvo <- data.frame(rgeoprofile:::boston_strangler)
test <- cgt_profile(desalvo$lat, desalvo$lon)
```

#### Mapping the Resultant Geographic Profile
Using the functions in the *rgeoprofile* package, the resultant geographic profile or jeopardy surface (three-dimensional probability surfaces which indicates the most probable area of offender anchor point) are returned for the input case and can then be mapped. The following will allow for the development of a *leaflet* map which plots the incidents and output from above:
```
g_map = sp::SpatialPixelsDataFrame(points = test[c("lons", "lats")], data = test)
g_map <- raster::raster(g_map)
# Assign a Coordinate Reference System for the Raster
raster::crs(g_map) <- sp::CRS("+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs")
# Define a Parula Color Pallete for Resultant Jeopardy Surface
library(leaflet) #for mapping the geographic profile
pal <- colorNumeric(pals::parula(200), raster::values(g_map),
    na.color = "transparent")
leaflet() %>%
    addTiles() %>%
    addProviderTiles('Esri.WorldTopoMap', group = 'Topo') %>%
    addAwesomeMarkers(lng = -71.07357, lat = 42.41322, icon =
        awesomeIcons(icon = 'home', markerColor = 'green'), popup = 'Residence') %>%
    addRasterImage(g_map, colors = pal, opacity = 0.6) %>%
    addLegend(pal = pal, values = raster::values(g_map), title = 'Score') %>%
    addCircleMarkers(lng = data$lon, lat = data$lat, radius = 4, opacity = 1,
        fill = 'black', stroke = TRUE, fillOpacity = 0.75, weight = 2,
        fillColor = "red")
```

![](../master/images/rossmo_geoprofile.png)

### neg_exp_profile
The *neg_exp_profile* function applies variations of the negative exponential decay model for serial crime analysis. In this model, the decline is at a constant rate, therefore the likelihood of the perpetrator's home base drops quickly from the incident locations until it approaches zero likelihood. The user can select different variants including the CrimeStat base model, the Dragnet model, or whether a buffer and plateau is present at the start of the decay function. This model assumes that the likelihood of the serial perpetrator's home base decreases in a exponential fashion as the distance increases from the crime incidents. This model utilizes Euclidean distance for the likelihood calculation.

An illustration of the difference between the base negative exponential function in comparison to the inclusion of a plateau and buffer is shown in the following:

![](../rgeoprofile/images/neg_exp_ddf.png)

```
test <- neg_exp_profile(desalvo$lat, desalvo$lon, method = "CrimeStat")
## See Mapping the Resultant Geographic Profile above for plotting the results
```

### Other Distance Decay Models
The following image provides a comparison of the different distance decay functions. Descriptions of the different functions will occur with usage examples thereafter. 

![](../master/images/crimestat_ddf.png)

#### linear_profile

The *linear_profile* function applies the linear decay model for serial crime analysis within CrimeStat. This model assumes that the likelihood of the serial perpetrator's home base decreases in a linear fashion as the distance increases from the crime incidents. This model utilizes Euclidean distance for the likelihood calculation. 

```
test <- linear_profile(desalvo$lat, desalvo$lon)
## See Mapping the Resultant Geographic Profile above for plotting the results
```

#### lognorm_profile

The *lognorm_profile* function applies the lognormal decay model for serial crime analysis within CrimeStat. This model is very similar to the normal model except with more skew to either side. If there is reason to believe that the perpetrator's residence is closer to the incidents, this function can take the form of a very rapid increase near incident with a gradual decline from the peak likelihood. This model utilizes Euclidean distance for the likelihood calculation.

```
test <- lognorm_profile(desalvo$lat, desalvo$lon)
## See Mapping the Resultant Geographic Profile above for plotting the results
```

#### norm_profile

The *norm_profile* function applies the normal decay model for serial crime analysis within CrimeStat. This model assumes that there is a peak likelihood of the serial perpetrator's home base at some optimal distance from the crime incidents. The function rises in likelihood to that distance and then declines at an equal rate (both prior to and after the peak likelhihood) giving the symetrical normal distribution. This model utilizes Euclidean distance for the likelihood calculation. 

```
test <- norm_profile(desalvo$lat, desalvo$lon)
## See Mapping the Resultant Geographic Profile above for plotting the results
```

#### trun_neg_exp_profile

The *trun_neg_exp_profile* function applies the truncated negative exponential decay model for serial crime analysis within CrimeStat. This is a joint function composed of both the linear and the negative exponential. For distances proximal to the incidents, a positive linear function is defined from zero likelihood at distance zero to a location of peak likelihood. At the peak likelihood the function takes the form of a negative exponential, rapidly declining as distance increases. This model utilizes Euclidean distance for the likelihood calculation. 

```
test <- trun_neg_exp_profile(desalvo$lat, desalvo$lon)
## See Mapping the Resultant Geographic Profile above for plotting the results
```

## Centrogrpahic Model Usage

The following examples illustrate the usage of the various centrographic functions within the **rgeoprofile** package.

### cmd_pred

The *cmd_pred* function calculates the center of minimum distance (CMD) for a set of incident coordinates. This function returns a singular coordinate for the CMD which can be used for geographic profiling or area prioritization under the assumption that the home base of the perpetrator is centrally located among the incidents.

```
cmd_pred(desalvo$lat, desalvo$lon)
```

### geom_mean_pred

The *geom_mean_pred* function calculates the geometric mean for a set of incident coordinates. This function returns a singular coordinate which can be used for geographic profiling or area prioritization under the assumption that the home base of the perpetrator is centrally located among the incidents.

```
geom_mean_pred(desalvo$lat, desalvo$lon)
```

### harm_mean_pred

The *harm_mean_pred* function calculates the harmonic mean for a set of incident coordinates. This function returns a singular coordinate which can be used for geographic profiling or area prioritization under the assumption that the home base of the perpetrator is centrally located among the incidents.

```
harm_mean_pred(desalvo$lat, desalvo$lon)
```

## Authors

* **Jamie Spaulding** - *Author, Contributor*

* **Keith Morris** - *Contributor*

## References
D Canter, T Coffey, M Huntley & C Missen. (2000). *Predicting serial killers' 
    home base using a decision support system.* Journal of quantitative 
    criminology, 16(4), 457-478.
    
Ned Levine, *CrimeStat IV: A Spatial Statistics Program for the Analysis 
    of Crime Incident Locations (version 4.0)*. Ned Levine & Associates,
    Houston, TX, and the National Institute of Justice, Washington, DC, June 2013.
    
DK Rossmo (1995). *Geographic profiling: Target patterns of serial 
    murderers.* Diss. Theses (School of Criminology)/Simon Fraser University.

DK Rossmo (2000). *Geographic profiling*. Boca Raton, FL: CRC Press.}


NEWS.md
# rgeoprofile (Development Version)

* Development version has no changes from 0.2.2.

# rgeoprofile 0.2.2

* Added `distances()` internal function to support the `cmd_pred()` function since the `aspace` package is scheduled for archival.

# rgeoprofile 0.2.1

* Fixed typo `cgt_profile()` function description.

* Adjustments made to the `circle_center()` function based on changes to the spatstat package architecture. The function is now solely dependent on the spatstat.geom sub-package.

* Updated maintainer contact information/affiliation.

# rgeoprofile 0.2.0

* Added `circle_center()` function to compute a center of the circle centrographic geographic profile prediction.

* Added /vignettes with examples of package functionality.

* Added NEWS.md for entire package history.

* Added hex sticker for package.

# rgeoprofile 0.1.1

* Change nearest neighbor calculation from orphaned RANN.L1 package to RANN package

# rgeoprofile 0.1.0

* Initial CRAN Release Version. 

FUNCTIONS IN INITAL VERSION

* `cgt_profile()`

* `cmd_pred()`

* `geom_mean_pred()`

* `harm_mean_pred()`

* `linear_profile()`

* `lognorm_profile()`

* `neg_exp_profile()`

* `norm_profile()`

* `trun_neg_exp_profile()`


R/neg_exp_profile.R
## neg_exp_profile
## Jamie Spaulding

#' Negative Exponential Model for Geographic Profiling
#' @description An implementation of variations of the negative exponential
#'     decay model for serial crime analysis. In this model, the decline is at
#'     a constant rate, therefore the likelihood of the perpetrator's home base
#'     drops quickly from the incident locations until it approaches zero
#'     likelihood. The user can select different variants including the 'CrimeStat'
#'     base model, the 'Dragnet' model, or whether a buffer and plateau is present
#'     at the start of the decay function. This model assumes that the likelihood
#'     of the serial perpetrator's home base decreases in a exponential fashion
#'     as the distance increases from the crime incidents.
#' @param lat a vector of latitudes for the crime incident series
#' @param lon a vector of latitudes for the crime incident series
#' @param method 'CrimeStat', 'Dragnet', or a custom parameter based negative exponential
#'     decay function. If using the 'CrimeStat' or 'Dragnet' method, values do not
#'     need to be provided from 'a' and 'b' as the default parameters will be
#'     used. Default parameters for the 'CrimeStat' are: \eqn{a = 1.89} \eqn{a = -0.06}.
#'     Default parameters for the 'Dragnet' are: \eqn{a = b = 1}. If using a custom
#'     model, values must be provided for '*a*' and '*b*'.
#' @param buffer TRUE/FALSE. Whether a buffer zone where a likelihood of zero
#'     is fit around the incidents and a plateau of peak likelihood is fit prior
#'     to the negative exponential decay. The function calculates the buffer zone
#'     and the plateau area to each be half of the average nearest neighbor
#'     distance.
#' @param a the slope coefficient which defines the function decrease in distance
#' @param b exponential multiplier for the distance decay function
#' @return A data frame of points depicting a spatial grid of the hunting area
#'     for the given incident locations. Also given are the resultant summed
#'     values (score) for each map point. A higher resultant score indicates
#'     a greater the probability that point contains the offender's anchor point.
#' @param n total number of cells within the spatial grid for the jeopardy surface.
#'     If \code{NULL}, the default value for '*n*' is 40,000.
#' @author Jamie Spaulding, Keith Morris
#' @references Ned Levine, \emph{CrimeStat IV: A Spatial Statistics Program for the
#'     Analysis of Crime Incident Locations (version 4.0)}. Ned Levine & Associates,
#'     Houston, TX, and the National Institute of Justice, Washington, DC, June 2013.
#' @references D Canter, T Coffey, M Huntley & C Missen. (2000). \emph{Predicting
#'     serial killers' home base using a decision support system.} Journal of
#'     quantitative criminology, 16(4), 457-478.
#' @keywords spatial methods
#' @examples
#' \dontshow{
#' data(desalvo)
#' test <- neg_exp_profile(desalvo$lat, desalvo$lon, method = "CrimeStat", n = 4)
#' }
#' \donttest{
#' #Using provided dataset for the Boston Strangler Incidents:
#' data(desalvo)
#' test <- neg_exp_profile(desalvo$lat, desalvo$lon, method = "CrimeStat")
#' g_map = sp::SpatialPixelsDataFrame(points = test[c("lons", "lats")], data = test)
#' g_map <- raster::raster(g_map)
#' # Assign a Coordinate Reference System for the Raster
#' raster::crs(g_map) <- sp::CRS("+proj=longlat +ellps=WGS84 +datum=WGS84 +no_defs")
#' # Define a Parula Color Pallete for Resultant Jeopardy Surface
#' library(leaflet) #for mapping the geographic profile
#' pal <- colorNumeric(pals::parula(200), raster::values(g_map),
#'     na.color = "transparent")
#' leaflet() %>%
#'     addTiles() %>%
#'     addProviderTiles('Esri.WorldTopoMap', group = 'Topo') %>%
#'     addAwesomeMarkers(lng = -71.07357, lat = 42.41322, icon =
#'         awesomeIcons(icon = 'home', markerColor = 'green'), popup = 'Residence') %>%
#'     addRasterImage(g_map, colors = pal, opacity = 0.6) %>%
#'     addLegend(pal = pal, values = raster::values(g_map), title = 'Score') %>%
#'     addCircleMarkers(lng = desalvo$lon, lat = desalvo$lat, radius = 4, opacity = 1,
#'         fill = 'black', stroke = TRUE, fillOpacity = 0.75, weight = 2,
#'         fillColor = "red")
#' }
#' @importFrom geosphere distHaversine
#' @importFrom RANN nn2
#' @importFrom utils txtProgressBar
#' @importFrom utils setTxtProgressBar
#' @export
neg_exp_profile <- function(lat, lon, method = c("CrimeStat", "Dragnet", "Custom"),
                            buffer = FALSE, a = NULL, b = NULL, n = NULL){
  # Set Defaults -----
  if (method == "Custom" & is.null(a)) {
    stop("If using a custom model, both 'a' and 'b' must be specified")
  }
  if (method == "Custom" & is.null(b)) {
    stop("If using a custom model, both 'a' and 'b' must be specified")
    }
  if (method == "CrimeStat") {
    a <- 1.89
    b <- -0.06
  }# Levine (2013)
  if (method == "Dragnet") {
    a <- 1
    b <- -1
  } #Canter et al. (2000)
  if (is.null(n)) {n <- 40000}

  # Computation of Map Boundaries/ Hunting Area -----
  lat_max <- max(lat) + ((max(lat) - min(lat)) / (2 * (length(lat) - 1)))
  lat_min <- min(lat) - ((max(lat) - min(lat)) / (2 * (length(lat) - 1)))
  lon_max <- max(lon) + ((max(lon) - min(lon)) / (2 * (length(lon) - 1)))
  lon_min <- min(lon) - ((max(lon) - min(lon)) / (2 * (length(lon) - 1)))

  # Calculate Range of Bounding Box -----
  lat_range <- lat_max - lat_min
  lon_range <- lon_max - lon_min

  # Determine Sequence of Lat and Lon Gridlines -----
  g_size <- sqrt(n)
  lats <- seq(lat_min,lat_max, length.out = g_size)
  lons <- seq(lon_min,lon_max, length.out = g_size)

  # Create a Run Sequence for Each Incident of Grid Points -----
  run_seq <- expand.grid(lats, lons)
  names(run_seq) <- c("lats", "lons")

  if (buffer == TRUE) {
    # Calculate Incident Buffer Zone -----
    dat_nn <- cbind(lat,lon) # Extract only lat and lon columns
    nn_list <- RANN::nn2(dat_nn, dat_nn, k=2) # Find NNs
    nn <- nn_list$nn.idx # Extract NN pairs

    # Calculate Distances Between NN Pairs -----
    nn_d <- NULL
    jj <- 1
    for(i in 1:nrow(nn)){
      incid1 <- dat_nn[nn[i,1],]
      incid2 <- dat_nn[nn[i,2],]
      nn_d[jj] <- geosphere::distHaversine(p1 = c(incid1[2], incid1[1]),
                                            p2 = c(incid2[2], incid2[1]),
                                            r = 3958) # hold y (lat) constant
      jj <- jj+1
    }
    plat_zone <- mean(nn_d) #Canter et al. (2000)
    buf_zone <- (mean(nn_d)) / 2 #Canter et al. (2000)

    jj <- 1
    output <- data.frame()
    # Progress Bar
    pb = utils::txtProgressBar(min = 0, max = length(lat) * n, style = 3)
    tick <- 0

    for(i in 1:length(lat)){
      for(j in 1:nrow(run_seq)){
        tick <- tick + 1
        utils::setTxtProgressBar(pb, tick)
        xn <- lon[i]
        yn <- lat[i]
        xi <- run_seq$lons[j]
        yi <- run_seq$lats[j]
        d <- geosphere::distHaversine(p1 = c(xn, yn),
                                      p2 = c(xi, yi),
                                      r = 3958)
        if(d < buf_zone) {out <- 0}
        if(d >= buf_zone & d < plat_zone) {out <- 1}
        if(d > plat_zone) {out <- (a * exp(b * (d - buf_zone)))}
        output[jj,i] <- out
        jj <- jj+1
      }
      jj <- 1
    }
    } else{
    jj <- 1
    output <- data.frame()
    pb = txtProgressBar(min = 0, max = length(lat) * n, style = 3)
    tick <- 0

    for(i in 1:length(lat)){
      for(j in 1:nrow(run_seq)){
        tick <- tick + 1
        utils::setTxtProgressBar(pb, tick)
        xn <- lon[i]
        yn <- lat[i]
        xi <- run_seq$lons[j]
        yi <- run_seq$lats[j]
        d <- geosphere::distHaversine(p1 = c(xn, yn),
                                      p2 = c(xi, yi),
                                      r = 3958)
        output[jj,i] <- a * exp(b * d)
        jj <- jj+1
      }
      jj <- 1
    }
    }

  # Summation of Values for Each Grid Point -----
  sums <- rowSums(output, na.rm = TRUE)
  dat <- cbind(sums, run_seq)
  return(dat)
}
